### ISSUE TITLE
Calibre B&N Cannot Retrieve Key

### ISSUE THREAD

**Issue by jokabomo** [apprenticeharper/DeDRM_tools#754]
I have installed the plugin in "Calibre 3.38.1". However, it cannot retrieve the key. I am sure the username/pass is correct. An error dialog pops up with:

> Could not retrieve key. Check username, password and intenet (sic) connectivity and try again.

To make sure the error was not on my part, I installed "Epubor Ultimate". That could also not login until I installed "Nook Study". With "Nook Study" running (and logged in ok), "Epubor" can decrypt files (but it is not free).

Not sure how to create a log. I ran "Calibre" from the command line with

`calibre-debug -g`

And there is nothing added to the window when I try to create a key. But here is the output:

>C:\Windows\System32>calibre-debug -g
calibre 3.38.1  embedded-python: True is64bit: False
Windows-10-10.0.17134-SP0 Windows ('32bit', 'WindowsPE')
32bit process running on 64bit windows
('Windows', '10', '10.0.17134')
Python 2.7.15+
Windows: ('10', '10.0.17134', 'SP0', u'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (6, 6, 1)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 81.5518394649 x 81.6428571429
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[0.14] splash screen shown
[0.14] Initializing db...
[0.16] db initialized
[0.16] Constructing main UI...
[1.14] main UI initialized...
[1.14] Hiding splash screen
[1.91] splash screen hidden
[1.91] Started up in 1.91 seconds with 3 books

**Comment by ElleKayEm** [comment]
Yes, B&N has made a recent change that seems to have completely cut support for Nook for PC off.  The B&N key retrieval in the plugin uses the same method that Nook for PC used so it is broken too.

However, Nookstudy for PC is still working.  DeDRM can retrieve the key from Nookstudy (as long as it is installed on the same machine and user as DeDRM) without the need to add your B&N account info to the plugin.

But who knows how much longer Nookstudy will be able to connect.  So I recommend freeing any books you have from them and not buying any more.

http://images.barnesandnoble.com/PResources/download/ereader2/sc/BN/pc/bndr_setup_2.1.2.28770.exe

Epubor does nothing more than these tools.  Do not waste your money.

**Comment by jokabomo** [comment]
ElleKayEm: Thank you. I didn't realize the plugin would work without a key. Your method works great: I downloaded all my books with Nook Study. Then, without generating a key, imported the books into Calibre from the Nook Study download folder (`C:\Users\<my user name>\Documents\My Barnes & Noble eBooks\<my email address>`) and the DRM has been removed.

**Comment by BertMedley** [comment]
I downloaded NOOKStudy as indicated and typed in the same username (amail) and password I use for thr BN sitre.  After showing a cycling circle for a couple of minutes It takes me back to the login.  Any ideas?

**Comment by ElleKayEm** [comment]
DeDRM is able to grab your B&N key directly from Nookstudy's logs.

**Comment by stoneasher** [comment]
I still have completely another decent solution contribution for This problem , i maneuvered it by doing this way :

1- if you need to DeDRM manually by the batch(other than using calbire), download nook study as others said .
2- download ebooks one by one .
3- go to DeDRM tools\Other_Tools\DRM_Key_Scripts\Barnes_and_Noble_ePubs\
4- run ignoblekey.pyw and it will generate b64 key file to be used with the windows batch DeDRM
5- drag and drop the protected ebooks into \DeDRM_tools-6.6.1\DeDRM_tools-6.6.1\contrib\windows\DeDRM_App\DeDRM_Drop_Target.bat
5- That's it

it worked with success on each processed BN epub
:)

**Comment by ElleKayEm** [comment]
I wouldn't think steps 3 and 4 are necessary.

**Comment by portalley** [comment]
Just fyi it looks like Nook Study finally stopped working at the beginning of March, so that leaves us without a workaround for this issue.

**Comment by ElleKayEm** [comment]
Some have had success with the Windows 8/10 Nook app.

**Comment by grillod** [comment]
Nook Study app stopped working for me as well. So as far as we  know, the plugin for Calibre is just broken for good?

**Comment by ElleKayEm** [comment]
Some have been able to use the Windows 8/10 app.

With some effort you could probably use the Android app or a Nook device.

**Comment by grillod** [comment]
Yes youre right. That seems to be the case. I gave up after customization failed. But importing does work on some of my mew books bit not others. Thanks!

**Comment by LFenske** [comment]
I was able to use the Microsoft Windows 10 Nook app (from the Microsoft store) to download books, but my credit card changed and I can't find the new Nook key.  Are you saying that some have been able to use the Windows 8/10 app to download the key or just to download books?

**Comment by ElleKayEm** [comment]
Just to get the books.  But if you have a key that worked previously, it may still work.  It does for me.  I've had the same key for years despite changed passwords.  Ever since they stopped using credit card based keys.

**Comment by SuiriuS** [comment]
Is there no way to obtain a key now? Does downloading books through the win 10 app not have drm? The win 10 app isnt working for me, the books wont download at all, but thats a separate issue. I retrieved the epubs from the android app but without a key I cant scrub the drm on most of them.

**Comment by ElleKayEm** [comment]
You can find it on a rooted Android device with the Nook app or a rooted Nook.

**Comment by LFenske** [comment]
I finally acquired an Android device that I could root, went through much pain to root it (the easy way!), retrieved the latest Nook key, removed DRM from my test book, and bought and un-DRMed the rest of the books in the series.

I don't know why my key changed, but I can now verify that the new key really is different from any of my old keys.

**Comment by ElleKayEm** [comment]
I know my key never changed after they switched away from cc-based keys about five years ago -- at least, not yet.   But others have reported theirs have during that time.  I have also seen a couple people report that they had different keys for different apps.  No clue why.

**Comment by SuiriuS** [comment]
Alright thanks, I freed my ebooks finally! 
I was able to get the key from the android app on a rooted android app, the file I needed was cchashdata.db as per the instructions in #814 
Then just followed the instructions for the calibre plugin and done.

**Comment by bfollowell** [comment]
Like all of you, I ran into this problem just recently. I've used the same decryption key for years, but was recently forced by B&N to change my account password, which broke my key. No idea if changing my password back will fix things, but I doubt it. I'll give it a try though. Otherwise, I'll try to get my new key from my rooted Nook HD+.
